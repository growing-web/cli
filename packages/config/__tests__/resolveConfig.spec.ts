import type { ManifestConfig } from '@growing-web/web-builder-types'
import { test, describe, expect } from 'vitest'
import path from 'path'
import {
  resolveUserConfig,
  resolveManifestConfig,
  validateManifestConfig,
} from '../src'

describe('resolve UserConfig test.', () => {
  const testFiles = ['cjs', 'module-js', 'ts', 'json', 'mjs']

  const _resolveUserConfig = async (root: string, mode = 'development') => {
    const rootDir = path.join(__dirname, root)
    const config = await resolveUserConfig(
      {
        mode: mode,
        command: 'dev',
      },
      rootDir,
    )
    return config
  }

  testFiles.forEach((file) => {
    test(`can load .${file} file`, async () => {
      const config = await _resolveUserConfig(`./fixtures/config/${file}/`)
      expect(config).toEqual({ foo: 'foo' })
    })
  })

  test(`when the configuration file does not exist`, async () => {
    const config = await _resolveUserConfig(`./fixtures/config/un-exits/`)
    expect(config).toEqual({})
  })

  test(`when the configuration include wrong value`, async () => {
    try {
      await _resolveUserConfig(`./fixtures/config/throw/`)
    } catch (error: any) {
      expect(error.message).to.include(
        'or the return value type of the function needs to be an object type',
      )
    }
  })

  test(`normal when the return value is a function`, async () => {
    const config = await _resolveUserConfig(
      `./fixtures/config/function-param/`,
      'dev',
    )
    expect(config).toEqual({ mode: 'dev' })
  })
})

describe('resolve ManifestConfig test.', () => {
  const _resolveManifestConfig = async (root: string) => {
    const rootDir = path.join(__dirname, root)
    const config = await resolveManifestConfig(rootDir)
    return config
  }
  test(`manifest files can be parsed by default.`, async () => {
    const config = await _resolveManifestConfig(`./fixtures/manifest/resolve/`)
    expect(config.schemaVersion).toBe('1.0.0')
  })

  test(`when manifest file does not exist.`, async () => {
    const config = await _resolveManifestConfig(
      `./fixtures/manifest/non-exits/`,
    )
    expect(config).toEqual({})
  })
})

describe('validate manifestConfig test.', () => {
  /**
   * Catch validation errors for testing
   * @param type
   * @param config
   */
  const validateConfigBySchema = (type: string, config: ManifestConfig) => {
    let error: any
    try {
      validateManifestConfig(config)
    } catch (errorFromPlugin: any) {
      error = errorFromPlugin
    } finally {
      if (type === 'success') {
        expect(error).toBeUndefined()
      } else if (type === 'failure') {
        expect(() => {
          if (!error) {
            return
          }
          throw error
        }).toThrowErrorMatchingSnapshot()
      }
    }
  }

  const tests: Record<'success' | 'failure', ManifestConfig[]> = {
    success: [
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js' }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          { input: 'app.js' },
          { input: 'foo.js', output: { name: 'foo' } },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', publicPath: '/' }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', publicPath: '/' }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { formats: ['umd', 'cjs'] } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { name: 'MyLib' } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { dir: 'dir' } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { externals: ['vue'] } }],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: { assetFileNames: '[name].[hash].[ext]' },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          { input: 'app.js', output: { chunkFileNames: '[name]-[hash].js' } },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          { input: 'app.js', output: { entryFileNames: '[name].[format].js' } },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { sourcemap: true } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { declaration: true } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', output: { globals: { jquery: '$' } } }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: { banner: { header: 'header', footer: 'footer' } },
          },
        ],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 6060,
        },
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          host: 'localhost',
        },
      },
      {
        schemaVersion: '1.0.0',
        manifests: ['exports-manifest'],
        entries: [
          {
            input: 'app.js',
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          proxy: [
            {
              url: '/api',
              target: 'http://localhost:8080/api',
              pathRewrite: [
                {
                  regular: '^/api',
                  replacement: '',
                },
              ],
            },
          ],
        },
      },
      {
        $schema:
          'https://unpkg.com/@growing-web/web-builder@latest/web-project-manifest.json',
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'index.html',
            publicPath: '/',
            output: {
              name: 'index',
              dir: 'dist',
              externals: [],
              assetFileNames: 'assets/[name].[hash].[ext]',
              chunkFileNames: 'assets/[name]-[hash].js',
              entryFileNames: 'assets/[name]-[hash].js',
              sourcemap: true,
              declaration: true,
              meta: {
                umdName: 'MyLib',
              },
            },
          },
          {
            input: 'index.js',
            publicPath: '/',
            output: {
              name: 'sub',
              dir: 'dist',
              externals: ['jquery'],
              assetFileNames: '[name].[hash].[ext]',
              chunkFileNames: '[name]-[hash].js',
              entryFileNames: '[name].[format].js',
              formats: ['esm', 'system', 'umd'],
              sourcemap: true,
              declaration: true,
              globals: {
                jquery: '$',
              },
              meta: {
                umdName: 'MyLib',
              },
              banner: {
                header: '/* library version 1.0.0 */',
                footer: '/* follow me on Twitter! @growing-web */',
              },
            },
          },
        ],
        server: {
          port: 3000,
          host: 'http://www.test.dev',
          proxy: [
            {
              url: '/api',
              target: 'http://localhost:8080/api',
              pathRewrite: [
                {
                  regular: '^/api',
                  replacement: '',
                },
              ],
            },
          ],
        },
      },
    ],
    failure: [
      {
        schemaVersion: 1,
        entries: [{ input: 'app.js' }],
      },
      {
        schemaVersion: 'a.b.c',
        entries: [{ input: 'app.js' }],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{ input: 'app.js', publicPath: './' }],
      },
      {
        schemaVersion: '1.0.0',
        manifests: ['exports'],
        entries: [
          {
            input: 'app.js',
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [{}],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              formats: ['es'],
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              formats: 'esm',
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              externals: {
                123: 'Vue',
              },
            },
          },
        ],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              name: 123,
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              dir: 123,
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              sourcemap: 'inline',
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              assetFileName: 1,
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              chunkFileNames: 1,
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              entryFileName: 1,
            },
          },
        ],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              declaration: 'false',
            },
          },
        ],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              banner: 'false',
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              banner: {
                header: 1,
                footer: 2,
              },
            },
          },
        ],
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              globals: ['$'],
            },
          },
        ],
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: '3000',
        },
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 1111111,
        },
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 6060,
          host: true,
        },
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 6060,
          host: 100,
        },
      },

      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 6060,
          host: 100,
          proxy: {},
        },
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
          },
        ],
        server: {
          port: 6060,
          host: 100,
          proxy: [
            {
              url: 123,
              target: '123',
            },
          ],
        },
      },
      {
        schemaVersion: '1.0.0',
        entries: [
          {
            input: 'app.js',
            output: {
              name: 'index',
            },
          },
          {
            input: 'foo.js',
            output: {
              name: 'index',
            },
          },
        ],
      },
    ] as any,
  }
  for (const type of Object.keys(tests) as ['success', 'failure']) {
    for (const value of tests[type]) {
      test(`should ${
        type === 'success' ? 'successfully validate' : 'throw an error on'
      } the patterns option: ${JSON.stringify(value)}`, () => {
        validateConfigBySchema(type, value)
      })
    }
  }
})
