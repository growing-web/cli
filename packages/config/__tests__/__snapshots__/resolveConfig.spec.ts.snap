// Vitest Snapshot v1

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"0.0.0","entries":["app.js"],"externals":{"123":"Vue"}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest has an unknown property 'externals'. These properties are valid:
   object { schemaVersion, entries, $schema?, server? }
   -> web-project manifest field in growing-web.
 - Manifest.entries[0] should be an object:
   object { input, publicPath?, output? }"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":["app.js"],"externals":{"123":"Vue"}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest has an unknown property 'externals'. These properties are valid:
   object { schemaVersion, entries, $schema?, server? }
   -> web-project manifest field in growing-web.
 - Manifest.entries[0] should be an object:
   object { input, publicPath?, output? }"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"assetFileName":1}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.assetFileName should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"assetFileName":1}}]} 2`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.assetFileName should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"assetFileName":1}}]} 3`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.assetFileName should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"banner":"false"}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.banner should be an object:
   object { header?, footer? }"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"banner":{"header":1,"footer":2}}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.banner.header should be a string.
 - Manifest.entries[0].output.banner.footer should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"banner":{"header":1123}}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.banner.header should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"chunkFileName":1}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.chunkFileName should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"declaration":"false"}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.declaration should be a boolean."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"dir":123}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.dir should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"entryFileName":1}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.entryFileName should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"externals":{"123":"Vue"}}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.externals should be an array:
   [string, ...]"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"formats":"esm"}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.formats should be an array:
   [\\"cjs\\" | \\"esm\\" | \\"iife\\" | \\"umd\\" | \\"system\\", ...]"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"formats":["es"]}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.formats[0] should be one of these:
   \\"cjs\\" | \\"esm\\" | \\"iife\\" | \\"umd\\" | \\"system\\""
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"globals":["$"]}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.globals should be an object:
   object { … } (additional property names should match pattern \\"[a-zA-Z]+\\")"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"name":123}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.name should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","output":{"sourcemap":"inline"}}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0].output.sourcemap should be a boolean."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js","publicPath":"./"}]} 1`] = `"entry.publicPath property must start with a /. Received: ./"`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":"3000"}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.port should be a number (should be >= 0 and <= 65535)."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":6060,"host":100,"proxy":[{"url":123,"target":"123"}]}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string.
 - Manifest.server.proxy[0].url should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":6060,"host":100,"proxy":[{"url":123}]}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string.
 - Manifest.server.proxy[0] misses the property 'target'. Should be:
   string
 - Manifest.server.proxy[0].url should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":6060,"host":100,"proxy":{}}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string.
 - Manifest.server.proxy should be an array:
   [object { url, target, secure?, changeOrigin?, pathRewrite?, … }, ...]"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":6060,"host":100}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":6060,"host":true}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.host should be a string."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{"input":"app.js"}],"server":{"port":1111111}} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.server.port should be >= 0 and <= 65535."
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"1.0.0","entries":[{}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.entries[0] misses the property 'input'. Should be:
   string"
`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":"a.b.c","entries":[{"input":"app.js"}]} 1`] = `"schemaVersion property must to comply with semver specification. Received: a.b.c"`;

exports[`validate manifestConfig test. > should throw an error on the patterns option: {"schemaVersion":1,"entries":[{"input":"app.js"}]} 1`] = `
"Invalid Manifest object. Web Project has been initialized using a Manifest object that does not match the API schema.
 - Manifest.schemaVersion should be a string.
   -> web project manifest version."
`;
